### PR TITLE
Added support for Single and Completable responses

### LINKED ISSUE #33: Update to RxSwift 3.4
Add support for Complete and Single observables

### PR DESCRIPTION
**Pull request by m-revetria**
Changes proposed:

* Support for new `Single` and `Completable` sequence types from `RxSwift`. Fixes #33 
* **Breaking change**: the return type of `var PaginationRequestType.collection` has been changed from `Observable<Base.Response>` to `Single<Base.Response>`* Old functions using `Observable` has been deprecated.
* Fixed compilation warnings

-- EDITED --

**Added breaking changes**: all previous rx functions returning an `Observable` has been deleted in favor of their `Single` and `Completable` versions

### CODE REVIEW

File: Sources/Common/RxManager.swift [hunk 1]
```diff
@@ -6,20 +6,252 @@
...
+        }
+    }
+
+    /**
+     Returns an `Single` of `OperaResult` for the current request. If something goes wrong an `OperaSwift.Error` error is propagated through the result sequence.
+
+     - returns: An instance of `Single<OperaResult>`
+     */
+    func response(_ requestConvertible: URLRequestConvertible) -> Single<OperaResult> {
+        return base.response(requestConvertible)
```
Comment (mcousillas6) [inline Sources/Common/RxManager.swift:83]: 🎉

File: Sources/Common/RxManager.swift [hunk 2]
```diff
@@ -6,20 +6,252 @@
...
+        return Single.just(json as Any)
+    }
+
+    /**
+     Returns an empty `Completable` sequence for the current request. No networking call is done as this is supposed to be used as part of a mocked response
+
+     - returns: An instance of `Completable`
+     */
+    func sampleCompletableResponse(_ route: RouteType) -> Completable {
+        return Completable.empty()
```
Comment (mcousillas6) [inline Sources/Common/RxManager.swift:163]: Does this work? If it is like Observable.empty() this will never emit to the subscriber

Comment (mcousillas6) [reply Sources/Common/RxManager.swift:163]: My bad, this will call onComplete.

**Review by mcousillas6 — APPROVED** [review]

File: Sources/Common/Error.swift [hunk 3]
```diff
@@ -181,18 +176,15 @@ extension OperaError: CustomStringConvertible {
...
         case .networking(let error, let request, _, _):
             var components = [String]()
-            if let error = error as? CustomStringConvertible {
-                components.append("Networking Error: \(error.description)")
-            }
-            if let error = error as? CustomDebugStringConvertible {
-                components.append(error.debugDescription)
-            }
+            components.append("Networking Error: \((error as CustomStringConvertible).description)")
+
```
Comment (dernster) [inline Sources/Common/Error.swift:181]: Isn't `components.append("Networking Error: \((error as CustomStringConvertible).debugDescription)")` missing?

File: Sources/Common/RouteType+Rx.swift [hunk 4]
```diff
@@ -23,52 +23,65 @@
...
+    /**
+     Returns a `Completable` sequence for the current request. If something goes wrong a Opera.Error error is propagated through the result sequence.
+
+     - returns: An instance of `Completable`
+     */
+    public func completable() -> Completable {
+        if base.manager.useMockedData && base.mockedData != nil {
+            return (base.manager as! RxManager).rx.completableResponse(base)
+        }
+        return (base.manager as! RxManager).rx.sampleCompletableResponse(base)
```
Comment (dernster) [inline Sources/Common/RouteType+Rx.swift]: `sampleCompletableResponse` and `completableResponse` are swapped?

**Review by dernster — CHANGES_REQUESTED** [review]

File: Sources/Common/RouteType+Rx.swift (hunk 4 above)
Comment (m-revetria) [reply Sources/Common/RouteType+Rx.swift]: looks like, yes. I'll take care of it

File: Sources/Common/Error.swift (hunk 3 above)
Comment (m-revetria) [reply Sources/Common/Error.swift:181]: Yes I noticed it, but as this is the implementation of the `description` property I think it is ok. You can check that there are other minor differences with the implementation of `debugDescroption` like the JSON value which is not included in the error's `description`. What do you think?

**Review by dernster — APPROVED** [review]

**Pull request merged**